src/wamr/filesystem.cpp

```diff
@@ -1,20 +1,432 @@
+    statApp->fs_rights_base = fileDesc.getActualRightsBase();
+    statApp->fs_rights_inheriting = fileDesc.getActualRightsInheriting();
+
+    // TODO - set fs flags
+    statApp->fs_flags = 0;
+
+    return __WASI_ESUCCESS;
+}
+
+static int32_t wasi_fd_fdstat_set_flags(wasm_exec_env_t exec_env,
+                                        int32_t a,
+                                        int32_t b)
+{
+    SPDLOG_DEBUG("S - fd_fdstat_set_flags");
+    throw std::runtime_error("fd_fdstat_set_flags not implemented");
+    return 0;
+}
+
+static int32_t doFileStat(int32_t fd,
+                          const std::string& relativePath,
+                          __wasi_filestat_t* statApp)
+{
+    WAMRWasmModule* module = getExecutingWAMRModule();
+    storage::FileDescriptor& fileDesc =
+      module->getFileSystem().getFileDescriptor(fd);
+    storage::Stat statNative = fileDesc.stat(relativePath);
+    if (statNative.failed) {
+        SPDLOG_ERROR("Failed stat: {}", statNative.wasiErrno);
+        return statNative.wasiErrno;
+    }
+
+    VALIDATE_NATIVE_ADDR(statApp, sizeof(__wasi_filestat_t));
+    statApp->st_dev = statNative.st_dev;
+    statApp->st_ino = statNative.st_ino;
+    statApp->st_filetype = statNative.wasiFiletype;
+    statApp->st_nlink = statNative.st_nlink;
+    statApp->st_size = statNative.st_size;
+    statApp->st_atim = statNative.st_atim;
+    statApp->st_mtim = statNative.st_mtim;
+    statApp->st_ctim = statNative.st_ctim;
+
+    return __WASI_ESUCCESS;
+}
+
+static int32_t wasi_fd_filestat_get(wasm_exec_env_t exec_env,
+                                    int32_t fd,
+                                    __wasi_filestat_t* statApp)
+{
+    SPDLOG_DEBUG("S - fd_filestat_get {}", fd);
+
+    return doFileStat(fd, "", statApp);
+}
+
+static int32_t wasi_fd_prestat_dir_name(wasm_exec_env_t exec_env,
+                                        int32_t fd,
+                                        char* path,
+                                        int32_t* pathLen)
+{
+    SPDLOG_DEBUG("S - fd_prestat_dir_name {}", fd);
+
+    WAMRWasmModule* module = getExecutingWAMRModule();
+    if (!module->getFileSystem().fileDescriptorExists(fd)) {
+        return __WASI_EBADF;
+    }
+
+    storage::FileDescriptor& fileDesc =
+      module->getFileSystem().getFileDescriptor(fd);
+    const std::string& pathStr = fileDesc.getPath();
+
+    VALIDATE_NATIVE_ADDR(path, pathStr.size());
+    std::copy(pathStr.begin(), pathStr.end(), path);
+
+    return __WASI_ESUCCESS;
+}
+
+typedef struct wasi_prestat_app
+{
+    __wasi_preopentype_t pr_type;
+    uint32_t pr_name_len;
+} wasi_prestat_app_t;
+
+static int32_t wasi_fd_prestat_get(wasm_exec_env_t exec_env,
+                                   int32_t fd,
+                                   wasi_prestat_app_t* prestatApp)
+{
+    SPDLOG_DEBUG("S - fd_prestat_get {}", fd);
+
+    WAMRWasmModule* module = getExecutingWAMRModule();
+    if (!module->getFileSystem().fileDescriptorExists(fd)) {
```

Comment:
Same point re. calling `getFileSystem()` twice and not using `module` variable.

Reply:
I need the `module` variable here to call `module->validateNativeAddress`.